panic: sync: WaitGroup is reused before previous Wait has returned

Error from the extension:
```
panic: sync: WaitGroup is reused before previous Wait has returned
goroutine 1 [running]:
sync.(*WaitGroup).Wait(0xc0001a3f10)
/usr/local/go/src/sync/waitgroup.go:132 +0xa5
main.main()
main.go:197 +0x87e
```


We'll have to revisit our use of a waitgroup to allow the send to the apm server to complete before the extension is frozen. Added in https://github.com/elastic/apm-aws-lambda/pull/106

## Current reproduction results

### Repro 1 : `main` branch, Ruby Runtime, function [EmilyTestRepro](https://eu-central-1.console.aws.amazon.com/lambda/home?region=eu-central-1#/functions/EmilyTestRepro?tab=testing)

With a lambda function similar to the one used by Emily, the error is successfully reproduced. I added additional to the extension to monitor the behavior of `backgroundDataSendWg`, which is the `WaitGroup` triggering the panic. The complete [logs](https://eu-central-1.console.aws.amazon.com/cloudwatch/home?region=eu-central-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252FEmilyTestRepro) are available here. The most relevant excerpt is below : 
```
START RequestId: 757912f1-8f13-4a2e-bb3a-a28619dd6b28 Version: $LATEST
2022/02/24 13:26:40 546ab00f-33b1-4915-ad6b-bb8d45ca5f37
2022/02/24 13:26:40 Register response: {
	"functionName": "EmilyTestRepro",
	"functionVersion": "$LATEST",
	"handler": "lambda_function.lambda_handler"
}
2022/02/24 13:26:40 Could not read ELASTIC_APM_DATA_RECEIVER_TIMEOUT_SECONDS, defaulting to 15: strconv.Atoi: parsing "": invalid syntax
2022/02/24 13:26:40 Extension listening for apm data on :8200
LOGS	Name: extension-dev-linux-amd64	State: Subscribed	Types: [platform]
2022/02/24 13:26:40 Waiting for next event...
2022/02/24 13:26:40 Server listening for logs data from AWS Logs API on sandbox:1234
I, [2022-02-24T13:26:41.218545 #17]  INFO  -- : [4.4.0] Starting agent, reporting to http://localhost:8200EXTENSION	Name: extension-dev-linux-amd64	State: Ready	Events: [INVOKE,SHUTDOWN]
2022/02/24 13:26:41 Received event: {
	"eventType": "INVOKE",
	"deadlineMs": 1645709221255,
	"requestId": "757912f1-8f13-4a2e-bb3a-a28619dd6b28",
	"invokedFunctionArn": "arn:aws:lambda:eu-central-1:627286350134:function:EmilyTestRepro",
	"tracing": {
		"type": "X-Amzn-Trace-Id",
		"value": "Root=1-62178790-41bd206a7002d2993498774a;Parent=61bd78a6009931ae;Sampled=0"
	}
}
2022/02/24 13:26:41 Checking for agent data
2022/02/24 13:26:41 No agent data on buffer
2022/02/24 13:26:41 Received log event platform.logsSubscription
first transaction
2022/02/24 13:26:41 Received log event platform.start
2022/02/24 13:26:41 Received log event platform.extension
second transaction
third transaction
2022/02/24 13:26:45 Adding agent data to buffer to be sent to apm server
2022/02/24 13:26:45 ADDING TO WAITGROUP
2022/02/24 13:26:45 Adding agent data to buffer to be sent to apm server
2022/02/24 13:26:45 Received log event platform.runtimeDone
2022/02/24 13:26:45 Received runtimeDone event for this function invocation
2022/02/24 13:26:45 Received runtimeDone signal
2022/02/24 13:26:45 WAITGROUP WAITING
2022/02/24 13:26:45 APM server response body: 
2022/02/24 13:26:45 APM server response status code: 202
2022/02/24 13:26:45 WAITGROUP DONE
2022/02/24 13:26:45 ADDING TO WAITGROUP
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 1 [running]:
sync.(*WaitGroup).Wait(0xc0001a7f10)
	/usr/local/go/src/sync/waitgroup.go:132 +0xa5
main.main()
	/Users/jlvoiseux/Documents/github/apm/apm-aws-lambda/apm-lambda-extension/main.go:200 +0x8da
END RequestId: 757912f1-8f13-4a2e-bb3a-a28619dd6b28
REPORT RequestId: 757912f1-8f13-4a2e-bb3a-a28619dd6b28	Duration: 4558.98 ms	Billed Duration: 4559 ms	Memory Size: 128 MB	Max Memory Used: 80 MB	Init Duration: 715.76 ms	
RequestId: 757912f1-8f13-4a2e-bb3a-a28619dd6b28 Error: exit status 2
Extension.Crash
```
Another call to `backgroundDataSendWg.Add(1)` seems to be done before the Waitgroup ends its `wait`, which is illegal.

### Repro 2 : `main` branch, Python Runtime, function [lambda-sample-dev-jlvoiseux](https://eu-central-1.console.aws.amazon.com/lambda/home?region=eu-central-1#/functions/lambda-sample-dev-jlvoiseux?tab=code)

Here, I tried to manually instrument a Python lambda handler to get as close as possible to translating the original Ruby function. Everything seems to work fine, and I cannot reproduce the issue - the wait group behaves as expected. Full logs are available [here](https://eu-central-1.console.aws.amazon.com/cloudwatch/home?region=eu-central-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252Flambda-sample-dev-jlvoiseux)
```
START RequestId: 27fe85f4-7f49-4e82-9c77-ae9d4479d00b Version: $LATEST
2022/02/24 14:19:14 Received event: {
	"eventType": "INVOKE",
	"deadlineMs": 1645712374347,
	"requestId": "27fe85f4-7f49-4e82-9c77-ae9d4479d00b",
	"invokedFunctionArn": "arn:aws:lambda:eu-central-1:627286350134:function:lambda-sample-dev-jlvoiseux",
	"tracing": {
		"type": "X-Amzn-Trace-Id",
		"value": "Root=1-621793e2-75b7dcad54c5cdb6504dcbc7;Parent=71703f007d5f1ea3;Sampled=0"
	}
}
2022/02/24 14:19:14 Checking for agent data
2022/02/24 14:19:14 No agent data on buffer
2022/02/24 14:19:14 Received log event platform.end
2022/02/24 14:19:14 Received log event platform.report
2022/02/24 14:19:14 Received log event platform.start
first transaction
second transaction
third transaction
2022/02/24 14:19:18 Adding agent data to buffer to be sent to apm server
2022/02/24 14:19:18 ADDING TO WAITGROUP
2022/02/24 14:19:18 Received log event platform.runtimeDone
2022/02/24 14:19:18 Received runtimeDone event for this function invocation
2022/02/24 14:19:18 Received runtimeDone signal
2022/02/24 14:19:18 WAITGROUP WAITING
2022/02/24 14:19:18 APM server response body: 
2022/02/24 14:19:18 APM server response status code: 202
2022/02/24 14:19:18 WAITGROUP DONE
2022/02/24 14:19:18 WAITGROUP RELEASED MAIN THREAD
2022/02/24 14:19:18 Checking for agent data
2022/02/24 14:19:18 No agent data on buffer
2022/02/24 14:19:18 Waiting for next event...
2022/02/24 14:19:18 funcDone signal received, not processing any more agent data
END RequestId: 27fe85f4-7f49-4e82-9c77-ae9d4479d00b
REPORT RequestId: 27fe85f4-7f49-4e82-9c77-ae9d4479d00b	Duration: 4175.15 ms	Billed Duration: 4176 ms	Memory Size: 128 MB	Max Memory Used: 67 MB	
```

@estolfo , when comparing your Ruby [handler](https://gist.github.com/jlvoiseux/3fae5fd3c8b3a843a588fe0900674908) with my Python [handler](https://gist.github.com/jlvoiseux/44910d1f8d061875345291304ff11c2c), do you observe any major difference ? If not, the issue might possibly be due to the way the Ruby agent sends its data ?

--- a/apm-lambda-extension/main.go
+++ b/apm-lambda-extension/main.go
@@ -135,19 +135,19 @@ func main() {
 			// Receive agent data as it comes in and post it to the APM server.
 			// Stop checking for, and sending agent data when the function invocation
 			// has completed, signaled via a channel.
+			backgroundDataSendWg.Add(1)
 			go func() {
+				defer backgroundDataSendWg.Done()
 				for {
 					select {
 					case <-funcDone:
 						log.Println("funcDone signal received, not processing any more agent data")
 						return
 					case agentData := <-agentDataChannel:
-						backgroundDataSendWg.Add(1)
 						err := extension.PostToApmServer(client, agentData, config)
 						if err != nil {
 							log.Printf("Error sending to APM server, skipping: %v", err)
 						}
-						backgroundDataSendWg.Done()
 					}
 				}
 			}()
@@ -194,13 +194,13 @@ func main() {
 				log.Println("Time expired waiting for agent signal or runtimeDone event")
 			}
 
+			close(funcDone)
 			backgroundDataSendWg.Wait()
 			if config.SendStrategy == extension.SyncFlush {
 				// Flush APM data now that the function invocation has completed
 				extension.FlushAPMData(client, agentDataChannel, config)
 			}
 
-			close(funcDone)
 			close(runtimeDoneSignal)
 			close(extension.AgentDoneSignal)
 		}

